### PR TITLE
Introduce native sockets

### PR DESCRIPTION
**Pull request by eugeneo**
This change reimplements the inspector server to be running in the same process on a separate thread.

It introduces basic websockets implementation written in C++.

Note that this change is not final, as I just discovered some crashes when running JS test suit.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,30 +18,22 @@ Pull inspector implementation as git submodule:
...
 
 The command will execute the script passed as the argument and start to wait for remote debugger
-connection on port 9222. To connect with DevTools, start Chromium with *--remote-debugging-targets=localhost:9222*
-and navigate to *chrome:inspect*. There is going to be a link to the Node debugger under "Devices > Remote Targets" section.
-
-Alternatively you could open the following URL directly:
+connection on port 5858. To connect with DevTools, start Chromium and open the following URL directly:
 ```text
-chrome-devtools://devtools/bundled/inspector.html?ws=localhost:9222/node
+chrome-devtools://devtools/bundled/inspector.html?ws=localhost:5858/node
```
Comment (repenaxa) [inline README.md]: Lets keep 9222 - it is easier to get docs when searching for remote and 9222.

File: configure [hunk 2]
```diff
@@ -373,6 +373,17 @@ parser.add_option('--enable-static',
     dest='enable_static',
     help='build as static library')
 
+parser.add_option('--no-browser-globals',
```
Comment (repenaxa) [inline configure]: Poor rebaseline?

File: README.md (hunk 1 above)
Comment (eugeneo) [inline README.md]: I'm using their default port that is read from their flags.

File: node.gyp [hunk 3]
```diff
@@ -719,7 +715,9 @@
         'GTEST_DONT_DEFINE_ASSERT_NE=1',
       ],
       'sources': [
+        'src/inspector_socket.cc',
```
Comment (repenaxa) [inline node.gyp]: Someone familiar with the target structure should say if it makes sense.

File: src/inspector_agent.cc [hunk 4]
```diff
@@ -0,0 +1,281 @@
+// Copyright Fedor Indutny and other Node contributors.
```
Comment (repenaxa) [inline src/inspector_agent.cc]: Is this the right copyright to use here?

File: src/inspector_agent.cc [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    buf->base = reinterpret_cast<char*>(malloc(len));
+    CHECK_NE(buf->base, nullptr);
+  }
+  buf->len = len;
+}
+
+class DispatchOnInspectorBackendTask : public v8::Task {
+public:
+  DispatchOnInspectorBackendTask(Agent* agent, const WTF::String& message) :
+                                                          agent_(agent),
```
Comment (repenaxa) [inline src/inspector_agent.cc]: Lets check with the node style on these offsets.

File: src/inspector_agent.cc [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+static double CurrentTime() {
+  return 0.0;
+}
+
+static void DisposeAsyncCb(uv_handle_t* handle) {
+  free(handle);
+}
+
+static void DisposeInspector(inspector_socket_t* socket, int status) {
```
Comment (repenaxa) [inline src/inspector_agent.cc]: I wonder if we can limit the C style to uv usage.

File: src/inspector_agent.cc [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+  if (len > 0) {
+    buf->base = reinterpret_cast<char*>(malloc(len));
+    CHECK_NE(buf->base, nullptr);
+  }
+  buf->len = len;
+}
+
+class DispatchOnInspectorBackendTask : public v8::Task {
+public:
+  DispatchOnInspectorBackendTask(Agent* agent, const WTF::String& message) :
```
Comment (repenaxa) [inline src/inspector_agent.cc]: You should use protocol::String16 for strings, WTF will go.

File: src/inspector_agent.cc [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+                                                          agent_(agent),
+                                                          message_(message) {}
+
+  void Run() override {
+    agent_->inspector_->dispatchMessageFromFrontend(message_);
```
Comment (repenaxa) [inline src/inspector_agent.cc]: We want to use the debugger queue for dispatching messages.

File: src/inspector_agent.cc [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+  }
+
+  virtual void sendProtocolNotification(PassOwnPtr<DictionaryValue> message) override {
+    sendMessageToFrontend(message);
+  }
+
+  virtual void flush() override { }
+
+  void sendMessageToFrontend(PassOwnPtr<DictionaryValue> message) {
+    WTF::String message_string = message->toJSONString();
```
Comment (repenaxa) [inline src/inspector_agent.cc]: String16

File: src/inspector_agent.cc [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+    agent_->write(message_string.utf8().data(), message_string.length());
```
Comment (repenaxa) [inline src/inspector_agent.cc]: You should perform node-side UTF conversion of the String16 here

File: src/inspector_agent.cc [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+  }
+
+  Agent* const agent_;
+};
+
+Agent::Agent(Environment* env) : port_(5858),
```
Comment (repenaxa) [inline src/inspector_agent.cc]: Agent term is overloaded - lets get a new name for this one. If it has a port, I consider it a "server"

File: src/inspector_agent.cc [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+                                 parent_env_(env),
+                                 platform_(nullptr) {
+  int err;
+
+  err = uv_sem_init(&start_sem_, 0);
+  CHECK_EQ(err, 0);
+}
+
+bool Agent::Start(v8::Platform* platform, int port, bool wait) {
+  WTF::Partitions::initialize(nullptr);
```
Comment (repenaxa) [inline src/inspector_agent.cc]: These will go along with WTF

File: src/inspector_agent.cc [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+  inspector_ = new blink::V8Inspector(env->isolate(), env->context(), platform);
+  inspector_->connectFrontend(new ChannelImpl(this));
+
+  int err;
+
+  platform_ = platform;
+
+  err = uv_loop_init(&child_loop_);
+  if (err != 0)
+    goto loop_init_failed;
```
Comment (repenaxa) [inline src/inspector_agent.cc]: Lets check the node style for goto, it might be discouraged.

File: src/inspector_agent.cc [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+  delete inspector_;
+}
+
+void Agent::OnRemoteData(uv_stream_t* stream, ssize_t read, const uv_buf_t* b) {
+
+  inspector_socket_t* socket =
+      reinterpret_cast<inspector_socket_t*>(stream->data);
+  Agent* agent = reinterpret_cast<Agent*>(socket->data);
+  if (read > 0) {
+    WTF::String message = WTF::String::fromUTF8(b->base, read - 1);
```
Comment (repenaxa) [inline src/inspector_agent.cc]: utf8 ditto

File: src/inspector_agent.cc [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+  uv_run(&child_loop_, UV_RUN_NOWAIT);
+
+  err = uv_loop_close(&child_loop_);
+  CHECK_EQ(err, 0);
+  delete inspector_;
+}
+
+void Agent::OnRemoteData(uv_stream_t* stream, ssize_t read, const uv_buf_t* b) {
+
+  inspector_socket_t* socket =
```
Comment (repenaxa) [inline src/inspector_agent.cc]: Do we have to use handle here - two interprets seems like an overkill.

File: src/inspector_agent.cc [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+  struct write_async_req* req = (struct write_async_req*) malloc(sizeof(struct write_async_req));
+  ASSERT_NE(req, nullptr);
+  req->data = (char*) malloc(len);
+  ASSERT_NE(req->data, nullptr);
+  memcpy(req->data, data, len);
+  req->agent = this;
+  req->len = len;
+  uv_async_init(&child_loop_, async, WriteCb);
+  async->data = req;
+  uv_async_send(async);
```
Comment (repenaxa) [inline src/inspector_agent.cc]: Is this passed with the ownership? Is this blocking on the other thread? Our messages can be 100Mb, you might want to send in chunks.

File: src/inspector_agent.cc [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+    }
+  }
+}
+
+void Agent::WorkerRun() {
+  int err;
+  sockaddr_in addr;
+  server.data = this;
+  uv_tcp_init(&child_loop_, &server);
+  uv_ip4_addr("0.0.0.0", port_, &addr);
```
Comment (repenaxa) [inline src/inspector_agent.cc]: bind on 127.0.0.1, also you might want to try binding on ipv6 ::1 if this one fails.

File: src/inspector_agent.h [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+class Environment;
+}  // namespace node
+
+namespace node {
+namespace inspector {
+
+class ChannelImpl;
+class DispatchOnInspectorBackendTask;
+
+class Agent {
```
Comment (repenaxa) [inline src/inspector_agent.h]: We should make sure the file names match the class names (or whatever policy node is using)

File: src/inspector_socket.cc [hunk 19]
```diff
@@ -0,0 +1,699 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#include "inspector_socket.h"
```
Comment (repenaxa) [inline src/inspector_socket.cc]: This is rather a web_socket, there is nothing inspector-specific here.

File: src/node.cc [hunk 20]
```diff
@@ -3264,15 +3259,17 @@ static bool ParseDebugOpt(const char* arg) {
     port = arg + sizeof("--debug-brk=") - 1;
   } else if (!strncmp(arg, "--debug-port=", sizeof("--debug-port=") - 1)) {
     port = arg + sizeof("--debug-port=") - 1;
-  } else if (!strncmp(arg, "--remote-debugging-port=",
```
Comment (repenaxa) [inline src/node.cc]: We should keep these flags

File: src/inspector_agent.cc (hunk 13 above)
Comment (eugeneo) [inline src/inspector_agent.cc]: This code was copied from the debug_agent.cc

File: configure (hunk 2 above)
Comment (eugeneo) [inline configure]: Fixed

File: src/inspector_agent.cc (hunk 5 above)
Comment (ofrobots) [inline src/inspector_agent.cc]: Node style would be to align  with `Agent*` on the first line.

File: test/cctest/inspector_socket.cc [hunk 21]
```diff
@@ -0,0 +1,593 @@
...
+
+struct read_expects {
+  const char* expected;
+  size_t expected_len;
+  size_t pos;
+  bool read_expected;
+};
+
+static const char HANDSHAKE_REQ[] = "GET / HTTP/1.1\r\n"
+                                    "Host: localhost:9222\r\n"
```
Comment (ofrobots) [inline test/cctest/inspector_socket.cc]: Should this be 5858? Can we avoid hard-coding it?

**Pull request closed without merging**